### PR TITLE
feat(contract): implement end_turn

### PR DESCRIPTION
**Pull request by rkdud007**
- [X] end_game _system
  - [X] When someone scored a goal, call this system
  - [X] Reset round and update player score
  - [X] A player has won 2 rounds the game ends
  - [X] Emit EndGame
- [X] end_game_test

### CODE REVIEW

**Comment by rkdud007** [comment]
currently end_game_test cause this error : https://github.com/dojoengine/dojo/issues/737

**Comment by LucasLvy** [comment]
can you format plz ?

File: contracts/src/events.cairo [hunk 1]
```diff
@@ -23,3 +23,9 @@ struct EndTurn {
     game_id: felt252,
     turn: felt252,
 }
+
+#[derive(Drop, Serde)]
+struct EndGame {
+    game_id: felt252,
+    round_number: u8,
```
Comment (LucasLvy) [inline contracts/src/events.cairo:30]: why do we want the round number here ? 
This should be added in the end turn

File: contracts/src/systems/end_game.cairo [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    fn execute(ctx: Context, game_id: felt252, scored_player: ContractAddress) {
+        let mut game = get!(ctx.world, game_id, Game);
+
+        // Reset round and update player score
+        game.turn = 0;
+        if (scored_player == game.player1) {
+            game.player1_score += 1;
+        } else if (scored_player == game.player2) {
+            game.player2_score += 1;
+        }
```
Comment (LucasLvy) [inline contracts/src/systems/end_game.cairo:18]: this should be in the end turn logic ? After all the cards have attacked you'll know if someone won the round or not

File: contracts/src/systems/end_game.cairo [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#[system]
+mod end_game_system {
+    use dojo::world::Context;
+    use starknet::ContractAddress;
+    use tsubasa::components::{Game, Outcome};
+    use tsubasa::events::{emit, EndGame};
+    use array::ArrayTrait;
+
+    fn execute(ctx: Context, game_id: felt252, scored_player: ContractAddress) {
```
Comment (LucasLvy) [inline contracts/src/systems/end_game.cairo:9]: Open question, do we call this system after every end turn and if the game shouldn't be ended it just does nothing or should it be called from the `end_turn` system if the conditions are met ?

File: contracts/src/events.cairo (hunk 1 above)
Comment (rkdud007) [reply contracts/src/events.cairo:30]: > The events log the round number
I understood this requirement as emit the current round number at the moment the game was end. Isn't 1 round contains multiple turns and 1 game contains multiple rounds?

**Comment by rkdud007** [comment]
@LucasLvy , Base on your comment I might confused about Round/Game/Turn terms. Here's my understanding--correct me if i'm wrong:

- **Turn**: A player takes an action (e.g., placing a card), computes defense, attacks, etc. Once completed, it's the other player's turn.

- **Round**: 1. A player can score a goal to end the round / 2. Ends when neither player has cards left; draw.

- **Game**: End when a player wins 2 rounds. Scores can be 0:2 or 1:2. ( no draw )

If this is correct, I would implement all in `end_turn` system:

**`end_turn` system**:
Purpose: Before ending a turn, check if conditions to end the round or game are met.
  1. Retrieve the current game status.
  2. Determine if the round should end (either due to a goal or a draw), and then update the round status.
  3. Check if the game should end and update the game status accordingly.
  
  + plus if then we need to Add `Round` component and modify `Game` component structure

**Comment by LucasLvy** [comment]
agree with everything you said, and indeed the `Draw` variant should be removed from the `Outcome` variant

**Comment by rkdud007** [comment]
If we need to track each status of Round also, we need Round as Component ( to store each outcome) . But if we do not need to track it, Then can just use original Game format :)

### Round 
```
struct Round {
    #[key]
    game_id: felt252,
    #[key]
    round_id: felt252,
     /// Current turn of the round.
    turn: felt252,
    /// Outcome of Round (player1/player2/draw)
    outcome: Option<Outcome>,
}
```

### Game
```
struct Game {
    /// Game id, computed as follows pedersen_hash(player1_address, player2_address) ??
    #[key]
    game_id: felt252,
    /// player 1 address.
    player1: ContractAddress,
    /// player 2 address.
    player2: ContractAddress,
    /// Rounds won by the player 1.
    player1_score: u8,
    /// Rounds won by the player 2.
    player2_score: u8,
     /// Rounds
    rounds : Array<Round>,
    /// Winner of the game. As long as it is `None` it means that the game is playing, cannot be Draw
    outcome: Option<Outcome>,
}
```

**Comment by LucasLvy** [comment]
I don't think we need to track them. There will be events that the frontend can listen to but we don't need details onchain the scores will be enough

File: contracts/src/events.cairo (hunk 1 above)
Comment (LucasLvy) [reply contracts/src/events.cairo:30]: It shouldn't be in the end game event at the end of the game. We don't care about the round number when the game is finished. There should probably be a EntRound event which would hold the round number and would be emitted in the `end_turn` system if a round is ending

**Comment by rkdud007** [comment]
### Decided to  wait until `attack` and `place_card` system to be implemented . Reason is : 

I would implement `turn_system` like this: 
```
 fn execute(ctx: Context, game_id: felt252, turn_calldata: TurnCalldata ){
        let game = get!(ctx.world, game_id, Game);
        // base on turn_calldata info decide attact/placecard
        if () {
        ctx.world.execute('attack_system', attack_calldata);
        else if () {
         ctx.world.execute('place_card_system', place_card_calldata);
         }

        // Switch to the other player's turn
        game.turn += 1,

        // emit EndTurn
        let mut values = ArrayTrait::new();
        serde::Serde::serialize(
            @EndTurn { game_id, turn: game.turn + 1, round_number: game.round_number }, ref values
        );
        emit(ctx, 'EndTurn', values.span());

        // Check round end conditions

        /// THIS IS THE CASE ROUND IS ENDED
        // if score goal 
        // Reset round and update player score
        game.turn = 0;
        if (scored_player == game.player1) {
            game.player1_score += 1;
        } else if (scored_player == game.player2) {
            game.player2_score += 1;
        }

        // emit EndTurn
        let mut values = ArrayTrait::new();
        serde::Serde::serialize(@EndRound { game_id, round_number: game.round_number }, ref values);
        emit(ctx, 'EndRound', values.span());

        /// THIS IS THE CASE GAME IS ENDED
        /// A player has won 2 rounds the game ends
        let mut game_outcome = game.outcome;
        if (game.player1_score == 2) {
            let winner = Outcome::Player1((game.player1));
            game_outcome = Option::Some(winner);
        } else if (game.player2_score == 2) {
            let winner = Outcome::Player2((game.player2));
            game_outcome = Option::Some(winner);
        }

        set!(
            ctx.world, Game {
                game_id: game_id,
                player1: game.player1,
                player2: game.player2,
                player1_score: game.player1_score,
                player2_score: game.player2_score,
                turn: game.turn,
                outcome: game_outcome,
            }
        );
    }
}
 ``` 
 
 where `turn_calldata` would define about which player chose to do which decision in this turn. 
 ### turn_calldata example
 ```
 enum Decision {
    /// some data needed to get information when execute attack 
    Attack,
    /// some data needed to get information when execute placecard 
    PlaceCard,
}

struct TurnCalldata {
    caller: ContractAddress,
    decision: Decision
}
```

So again, Base on that, would better to wait until `attack` and `place_card` system to implemented. Because end_turn need to imply logic for either attack/place_card during the turn and determine whether end round/game. -- turned into draft.

File: contracts/src/events.cairo [hunk 4]
```diff
@@ -25,3 +25,9 @@ struct CardPlaced {
     card_id: u256,
     position: Roles,
 }
+
+#[derive(Drop, Serde)]
+struct EndGame {
+    game_id: felt252,
+    round_number: u8,
```
Comment (LucasLvy) [inline contracts/src/events.cairo:32]:
```suggestion
```

File: contracts/src/systems/end_turn.cairo [hunk 5]
```diff
@@ -15,25 +15,31 @@ mod end_turn_system {
...
+        emit!(ctx.world, EndTurn { game_id, turn: game.turn })
+
+        // Increments the energy of the player 
         let mut player = get!(ctx.world, (game_id, ctx.origin), Player);
         player.remaining_energy = game.turn / 2 + 2;
         set!(ctx.world, (player));
 
-        emit!(ctx.world, EndTurn { game_id, turn: game.turn + 1 })
+        /// End the Game
+        /// If one reached score 2, set winner
```
Comment (LucasLvy) [inline contracts/src/systems/end_turn.cairo:32]:
```suggestion
        // End the Game
        // If one reached score 2, set winner 
```

File: contracts/src/systems/end_turn.cairo [hunk 6]
```diff
@@ -15,25 +15,31 @@ mod end_turn_system {
...
+        /// End the Game
+        /// If one reached score 2, set winner 
+        let mut game_outcome = game.outcome;
+        if (game.player1_score == 2) {
+            let winner = Outcome::Player1((game.player1));
+            game_outcome = Option::Some(winner);
+        } else if (game.player2_score == 2) {
+            let winner = Outcome::Player2((game.player2));
+            game_outcome = Option::Some(winner);
+        }
```
Comment (LucasLvy) [inline contracts/src/systems/end_turn.cairo:40]:
```suggestion
        game.outcome = Option::Some(if (game.player1_score == 2) {
            Outcome::Player1(game.player1)
        } else if (game.player2_score == 2) {
            Outcome::Player2(game.player2)
        });
```

**Review by LucasLvy — CHANGES_REQUESTED** [review]

**Review by LucasLvy — APPROVED** [review]

**Pull request merged**